packages/react-native/ReactCommon/react/renderer/attributedstring/AttributedString.cpp

```diff
@@ -91,6 +91,18 @@ void AttributedString::prependAttributedString(
       attributedString.fragments_.end());
 }
 
+int AttributedString::countAttachments() const {
+  int count = 0;
+
+  for (const auto& fragment : fragments_) {
+    if (fragment.isAttachment()) {
+      ++count;
+    }
+  }
+
+  return count;
+}
```

Comment:
As far as I can see, this is used only in TextLayoutManager. Why do we need to uplift this function into the `AttributedString` module and make it public? 

This will increase the public API surface and thus it will increase the maintenance burden.

Can we have it as a private function inside the `TextLayoutManager`, for the time being?
We can always export it to be a public function in a second moment, when we actually need it.

Reply:
> increase the maintenance burden

Will it? In what way, exactly? I'm sorry, but I can't see it.

Reply:
Actually, I didn't really mean to start a discussion here. The truth is that I might not have the mental resources for this. It's month 7 of my project and it's been tough. Contributing to React Native is really difficult. I know it's not your fault, but I just wanted to share.

Could we accept the burden risk of this one-loop method?

Reply:
no, i'd rather avoid. The maintenance burden is not there now but it can creep-in easily.
Say that another module needs something similar and hook into the AttributedString one. Now we have a dependency that we need to maintain and we can't break.
The whole point of keeping the methods private if they need to be used only by a single file is to prevent the creation of these new dependnecy that we might not want.

Reply:
I can't speak about "something similar", because that's theorizing, but if someone ever needs the attachment count of an `AttributedString` instance, it's the exact place where they should look for it.

> Now we have a dependency

...on a method? You're talking about "adding" and "maintaining" dependencies as if I my diff included any `#include` term.

> that we need to maintain

How does one maintain a method call?

> and we can't break

Why can't we remove a method call if it's not needed? If it's needed, why would we remove it?

Are you willing to leave the snippet duplicated, out of fear that we might use the method in a different place?

Reply:
nono, the deduplication is good. What I am suggesting is that:
1. we remove the code from AttributedString
2. we factor out the duplicated code in a method into the `TextLayoutManager.cpp`

There is no need to add a new definition in the `AttributedString.h`. The method does not exists there and I'd rather not add it:
1. We don't need it there
2. Once we put it there we can't remove it without a breaking change
3. People might start depending on it, and this will create a new dependency we can avoid for now.

Reply:
> Once we put it there we can't remove it without a breaking change

Maybe I'm missing something big here. Does `ReactCommon` C++ API have third-party dependents?

Reply:
> Just to be on the same page, are you aware that I'm about to add a major, user-perceived feature, with implementation both in this class and the whole React Native stack, intrusive on basically every possible level?

Yes, I'm aware. I read [through](https://github.com/react-native-community/discussions-and-proposals/issues/695). But it doesn't mean that we can go rogue and mess with the public API surface of React Native. We have efforts to try and reduce it in a meaningful way, to keep providing all the functionalities that are needed but in a controlled way. The goal is to reach a point when we can do React Native minor with no breaking changes.

The path is still long and hard, and we should be mindful when we modify header files. For the sake of this pull request, there is no need to expand the API surface of `AttributedString`, therefore, we should not do it.

If in a future change, you are going to need that method in a module that is not `TextLayoutManager` we will move it.

> Maybe I'm missing something big here. Does ReactCommon C++ API have third-party dependents?

Not sure what do you mean, but the answer is yes:
- there are third party libraries that use ReactCommon directly to access the C++ layer of ReactNative (A dependency with the direction `3rd party libs --> ReactCommon`)
- ReactCommon uses `glog`, `boost`, `double-conversion`, `folly`, `hermes` ... (A dependency with the direction `ReactCommon --> 3rd party libs`)

Reply:
I wrongly understood that `ReactCommon` is an implementation detail of `ReactAndroid` and React/iOS. When you were talking about dependencies, I considered them internal dependencies within the project.

Is every single header withing `ReactCommon` source tree considered a part of the external public API?

Reply:
Here are the [GitHub usages of this header](https://github.com/search?q=%23include+%3Creact%2Frenderer%2Fattributedstring%2FAttributedString.h%3E+NOT+path%3A**%2FReactCommon%2Freact%2F**%2F*+NOT+path%3A**%2F.pnpm-store%2F**+NOT+is%3Afork&type=code).

I excluded forks and quasi-forks.

Reply:
> Is every single header within ReactCommon source tree considered a part of the external public API?

Unfortunately yes, because we have never split them between private and public. This is something we would like to do in the close next future (this year or beginning of the next)

Reply:
Made a local helper function